### PR TITLE
Codechange: Clean up scrollbars by adding iteration

### PR DESCRIPTION
**Pull request by PeterN**
## Motivation / Problem

There are about 5 million different way we iterate through visible scrollbar elements. They are mostly very C-style, and often involve casts due to data type conversion issues.

## Description

Implement a range iterator to allow iteration through the visible elements that a scrollbar refers to.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

File: src/widget_type.h [hunk 1]
```diff
@@ -17,6 +17,25 @@
 #include "gfx_type.h"
 #include "window_type.h"
 
+/* Range Iterator Helper. This should be somewhere else, and documented. Where? */
+template <class Iter>
```
Comment (rubidium42) [inline src/widget_type.h]: Isn't this just a simple version of `(std::)span` (core/span_type.hpp)?

Comment (PeterN) [reply src/widget_type.h]: Perhaps, now trying to work out how to use that.

Comment (PeterN) [reply src/widget_type.h]: Unless I was trying it wrong, our span only seems to support iterators that are pointers.

**Comment by PeterN** [comment]
Mostly implemented by #12190

**Pull request closed without merging**